File: go/diffTime.go
package _go

import (
    "time"
    "fmt"
)
//换算之前某个时间为当前月份的时间
func GenStartTime(startStr string) (time.Time, error) {
    startTime, err := time.Parse("2006-01-02 15:04:05", startStr) //转化为：2006-01-02 15:04:05这个格式
    if err != nil {
	return startTime, err
    }
    now := time.Now()
    nY, nm, nd := now.Date()        //获取今天年月日的天数
    _, _, sd := startTime.Date()    //获取开始的日
    nH, ni, ns := now.Clock()       //获取今天的时分秒
    sH, si, ss := startTime.Clock() //获取开始时分秒
    startDay := sd
    ld := GetLastOfMonth(startTime)
    lnd := GetLastOfMonth(now)
    if sd == ld { //说明是月份最后一天,
	startDay = lnd
    }
    startTime = time.Date(nY, nm, startDay, sH, si, ss, 0, time.UTC)
    if (nd < sd) || (nd == sd && time.Date(2006, 01, 02, nH, ni, ns, 0, time.UTC).Before(time.Date(2006, 01, 02, sH, si, ss, 0, time.UTC))) {
	//(现在时间<开始时间）
	//如果现在时间<开始时间，开始时间减一个月
	startTime = OffMonths(startTime, 1)

    }
    return startTime, nil
}

//指定日期当月的最后一天
func GetLastOfMonth(now time.Time) int {
    currentYear, currentMonth, _ := now.Date()
    firstOfMonth := time.Date(currentYear, currentMonth, 1, 0, 0, 0, 0, time.UTC)
    lastOfMonth := firstOfMonth.AddDate(0, 1, -1)
    return lastOfMonth.Day()
}

//增加月份,处理月份天数的临界值
func AddMonths(timer time.Time, month time.Month) time.Time {
    Y, m, d := timer.Date()
    H, i, s := timer.Clock()
    newM := m
    if (m + month) > 12 {
	Y += int((m + month) / 12)
	newM = (m + month) % 12
    } else {
	newM = m + month
    }
    lastDay := d
    if GetLastOfMonth(timer) == d {
	newTime := time.Date(Y, newM, 1, 0, 0, 0, 0, time.UTC)
	lastDay = GetLastOfMonth(newTime)
    }
    return time.Date(Y, newM, lastDay, H, i, s, 0, time.UTC)
}

//减少月份处理月份天数的临界值
func OffMonths(timer time.Time, month time.Month) time.Time {
    Y, m, d := timer.Date()
    H, i, s := timer.Clock()
    newM := m
    if (m - month) <= 0 {
	year := int((m-month)/12) + 1
	Y = Y - year
	newM = m + time.Month(year*12) - month
    } else {
	newM = m - month
    }
    lastDay := d

    if GetLastOfMonth(timer) == d {
	newTime := time.Date(Y, newM, 1, 0, 0, 0, 0, time.UTC)
	lastDay = GetLastOfMonth(newTime)
    }
    return time.Date(Y, newM, lastDay, H, i, s, 0, time.UTC)
}

//计算两个日期之间的相差几年几个月几天
func DiffDate(dateOne, dateTwo time.Time) map[string]int {
    result := make(map[string]int)

    if dateOne == dateTwo {
	result["year"], result["month"], result["day"] = 0, 0, 0
	return result
    }
    if dateOne.Unix() > dateTwo.Unix() {
	tmp := dateTwo
	dateTwo = dateOne
	dateOne = tmp
    }
    fmt.Println(dateOne)
    Y1, m1, d1 := dateOne.Date()
    Y2, m2, d2 := dateTwo.Date()
    H1, i1, _ := dateOne.Clock()
    H2, i2, _ := dateTwo.Clock()
    Y := Y1 - Y2
    m := m2 - m1
    d := 0
    if GetLastOfMonth(dateOne) == d1 && GetLastOfMonth(dateTwo) == d2 {
	d = 0
    } else {
	d = d2 - d1
    }
    if H1 < H2 || (H1 == H2 && i1 < i2) {
	d++
    }
    if d < 0 {
	d += GetLastOfMonth(dateTwo)
	m--
    }

    if m < 0 {
	m += 12
	Y--
    }
    result["year"], result["month"], result["day"] = Y, int(m), d

    return result

}

File: php/diffTime.php
<?php

/**
 * 时间相关函数.
 */



class Time
{
    const FormatDateWithTime = 'Y-m-d H:i:s';

    const FormatDate = 'Y-m-d';

    const FormatUsDate = 'm/d/Y';
    /**
     * 计算两个日期之间的月份和天数差
     * @param $date1
     * @param $date2
     * @return array
     */
    public static function diffDate($dateTime1, $dateTime2)
    {
        if ($dateTime1 === $dateTime2) {
            return array('year' => 0, 'month' => 0, 'day' => 0);
        } else {
            if (strtotime($dateTime1) > strtotime($dateTime2)) {
                $tmp = $dateTime2;
                $dateTime2 = $dateTime1;
                $dateTime1 = $tmp;
            }
            $date1 = explode(" ", $dateTime1)[0];
            $date2 = explode(" ", $dateTime2)[0];
            list($Y1, $m1, $d1) = explode('-', $date1);
            list($Y2, $m2, $d2) = explode('-', $date2);

            $Y = $Y2 - $Y1;
            $m = $m2 - $m1;
            //判断两个时间是否是月的最后一天,,
            if (self::getMonthLastDay($date1) == $d1 && self::getMonthLastDay($date2) == $d2) {
                $d = 0;
            } else {
                $d = $d2 - $d1;
            }

            $time1 = explode(" ", $dateTime1)[1];
            $time2 = explode(" ", $dateTime2)[1];
            list($H1, $i1, $s1) = explode(':', $time1);
            list($H2, $i2, $s2) = explode(':', $time2);

            if ($H1 < $H2 || ($H1 == $H2 && $i1 < $i2)) {
                $d++;
            }

            if ($d < 0) {
                $d += (int)date('t', strtotime("-1 month $date2"));
                $m--;

            }
            if ($m < 0) {
                $m += 12;
                $Y--;
            }


            return array('year' => $Y, 'month' => $m, 'day' => $d);
        }

    }

    /**
     * 获取这个日期当月的最后一天
     * @param $date
     * @return false|string
     */
    public static function getMonthLastDay($date)
    {
        $firstday = date('Y-m-01', strtotime($date));
        return date('d', strtotime("$firstday +1 month -1 day"));
    }

    /**
     * 增加月份 对月份临界值处理
     * @param $date
     * @param $month
     * @return false|string
     */

    public static function addMonths($date, $month)
    {
        $dateArray = explode(" ", $date);
        $dateDay = $dateArray[0];
        if (count($dateArray) == 2) {
            $dateTime = $dateArray[1];
        } else {
            $dateTime = "00:00:00";
        }
        list($Y, $m, $d) = explode('-', $dateDay);
        list($H, $i, $s) = explode(':', $dateTime);
        if (($m + $month) > 12) {
            $Y += floor(($m + $month) / 12);
            $newM = ($m + $month) % 12;
        } else {
            $newM = $m + $month;
        }
        $lastDay = $d;
        if (self::getMonthLastDay($dateDay) == $d) {//说明是最后一天
            $newTime = mktime($H, $i, $s, $m, 5, $Y);
            //获取下个月最后一天
            $lastDay = self::getMonthLastDay(date('Y-m-d', $newTime));
        }
        $newDate = date("Y-m-d H:i:s", mktime($H, $i, $s, $newM, $lastDay, $Y));
        return $newDate;
    }

    /**
     * 减少月份,对月份临界值处理
     * @param $date
     * @param $month
     * @return false|string
     */
    public static function offMonths($date, $month)
    {
        $dateArray = explode(" ", $date);
        $dateDay = $dateArray[0];
        if (count($dateArray) == 2) {
            $dateTime = $dateArray[1];
        } else {
            $dateTime = "00:00:00";
        }

        list($Y, $m, $d) = explode('-', $dateDay);
        list($H, $i, $s) = explode(':', $dateTime);
        if (($m - $month) <= 0) {
            $year = ceil(($m - $month) / 12);
            $Y = $Y - $year;
            $newM = $m + $year * 12 - $month;
        } else {
            $newM = $m - $month;
        }
        $lastDay = $d;
        if (self::getMonthLastDay($dateDay) == $d) {//说明是最后一天
            $newTime = mktime($H, $i, $s, $m, 1, $Y);
            //获取下个月最后一天
            $lastDay = self::getMonthLastDay(date('Y-m-d', $newTime));
        }
        $newDate = date("Y-m-d H:i:s", mktime($H, $i, $s, $newM, $lastDay, $Y));
        return $newDate;
    }

    /**
     * 换算之前某个时间为当前月份的时间
     * @param $date
     * @return false|string
     */
    public static function genStartTime($date)
    {
        $dateArray = explode(" ", $date);
        $dateDay = $dateArray[0];
        if (count($dateArray) == 2) {
            $dateTime = $dateArray[1];
        } else {
            $dateTime = "00:00:00";
        }
        list($Y, $m, $d) = explode('-', $dateDay);
        list($H, $i, $s) = explode(':', $dateTime);

        $nowDate = self::format(time());
        $nowDateArray = explode(" ", $nowDate);
        $nowDay = $nowDateArray[0];
        $nowTime = $nowDateArray[1];
        list($nY, $nm, $nd) = explode('-', $nowDay);
        list($nH, $ni, $ns) = explode(':', $nowTime);


        $ld = self::getMonthLastDay($date);
        $lnd = self::getMonthLastDay($nowDate);

        $startDay = $d == $ld ? $lnd : $d;
        $startTime = self::format(mktime($H, $i, $s, $nm, $startDay, $nY));
        if (($d > $nd) || ($d == $nd && mktime($H, $i, $s) > mktime($nH, $ni, $ns))) {
            $startTime = self::offMonths($startTime, 1);
        }
        return $startTime;
    }
}
